enterprise/server/ociregistry/ociregistry.go

```diff
@@ -68,240 +67,150 @@ func (r *registry) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-		// This is an "image index", a meta-manifest that contains a list of
-		// {platform props, manifest hash} properties to allow client to decide
-		// which manifest they want to use based on platform.
-		case types.OCIImageIndex, types.DockerManifestList:
-			return nil, status.UnknownErrorf("could not get image in image index from descriptor: %s", err)
-		case types.OCIManifestSchema1, types.DockerManifestSchema2:
-			return nil, status.UnknownErrorf("could not get image from descriptor: %s", err)
-		default:
-			return nil, status.UnknownErrorf("descriptor has unknown media type %q, oci error: %s", remoteDesc.MediaType, err)
+		if err != context.Canceled {
+			log.CtxWarningf(ctx, "error writing response body for '%s', upstream '%s': %s", inreq.URL.String(), u.String(), err)
 		}
+		return
 	}
-	return img, nil
 }
 
-// handleBlobRequest fetches all or part of a blob from a remote registry.
-// Used to pull OCI image layers.
-func (r *registry) handleBlobRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, imageName, refName string) {
-	reqStartTime := time.Now()
-
-	d, err := name.NewDigest(imageName + "@" + refName)
-	if err != nil {
-		http.Error(w, fmt.Sprintf("invalid hash %q: %s", refName, err), http.StatusNotFound)
+func (r *registry) handleBlobsOrManifestsRequest(ctx context.Context, w http.ResponseWriter, inreq *http.Request, blobsOrManifests, repository, identifier string) {
+	if "blobs" == blobsOrManifests && !isDigest(identifier) {
+		http.Error(w, fmt.Sprintf("can only retrieve blobs by digest, received '%s'", identifier), http.StatusNotFound)
 		return
 	}
 
-	layer, err := r.getBlob(ctx, d)
+	ref, err := parseReference(repository, identifier)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("error fetching layer %q: %s", d, err), http.StatusNotFound)
+		http.Error(w, fmt.Sprintf("error parsing image repository '%s' and identifier '%s': %s", repository, identifier, err), http.StatusNotFound)
 		return
 	}
-	blobSize, err := layer.Size()
+
+	u := url.URL{
+		Scheme: ref.Context().Scheme(),
+		Host:   ref.Context().RegistryStr(),
+		Path:   "/v2/" + ref.Context().RepositoryStr() + "/" + blobsOrManifests + "/" + ref.Identifier(),
+	}
+	upreq, err := http.NewRequest(inreq.Method, u.String(), nil)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("error getting layer size %q: %s", d, err), http.StatusNotFound)
+		http.Error(w, fmt.Sprintf("could not make %s request to upstream registry '%s': %s", inreq.Method, u.String(), err), http.StatusNotFound)
 		return
 	}
-
-	w.Header().Set("Content-Length", strconv.FormatInt(blobSize, 10))
-
-	// If this is a HEAD request, and we have already figured out the blob
-	// length then we are done.
-	if req.Method == http.MethodHead {
-		w.WriteHeader(http.StatusOK)
-		return
+	if inreq.Header.Get(headerAccept) != "" {
+		upreq.Header.Set(headerAccept, inreq.Header.Get(headerAccept))
 	}
-
-	offset := int64(0)
-	limit := int64(0)
-
-	if r := req.Header.Get("Range"); r != "" {
```

Comment:
Instead of dropping this check entirely, can we instead return an "unsupported" error on a partial request?

Reply:
For sure. Returning StatusNotImplemented.